pypechain/render/contract.py

```diff
@@ -19,14 +23,98 @@
     get_structs_for_abi,
     is_abi_event,
     is_abi_function,
-    load_abi_from_file,
 )
 from pypechain.utilities.format import capitalize_first_letter_only
 from pypechain.utilities.templates import get_jinja_env
-from pypechain.utilities.types import EventData, FunctionData, SignatureData, gather_matching_types
+from pypechain.utilities.types import EventData, FunctionData, SignatureData
 
 
-def render_contract_file(contract_name: str, abi_file_path: Path) -> str:
+@dataclass
+class ContractInfo:
+    """Contract Information."""
+
+    abi: ABI
+    bytecode: str
+    contract_name: str
+    structs: dict[str, StructInfo]
+    events: dict[str, EventInfo]
+
+
+def get_contract_infos(abi_infos: list[AbiInfo]) -> dict[str, ContractInfo]:
+    """Get a dictionary of ContractInfos from an AbiInfos list.
+
+     This helps us organize structs and events by the contracts they are defined
+     in. Because structs can be defined in one file, and used in another, we
+     need to avoid duplication of the definitions.
+
+    Parameters
+    ----------
+    abi_infos : list[AbiInfo]
+        _description_
+
+    Returns
+    -------
+    dict[str, ContractInfo]
+        _description_
+    """
+    contract_infos: dict[str, ContractInfo] = {}
+
+    # Populate contract_infos with all the structs, events, whole ABIs, bytecodes and contract names.
+    for abi_info in abi_infos:
+        structs = get_structs_for_abi(abi_info.abi)
+        events = get_events_for_abi(abi_info.abi)
+        contract_infos[abi_info.contract_name] = ContractInfo(
+            abi=abi_info.abi, bytecode=abi_info.bytecode, contract_name=abi_info.contract_name, structs={}, events={}
+        )
+        add_structs(contract_infos, structs)
+        add_events(contract_infos, events, abi_info.contract_name)
+
+    return contract_infos
+
+
+def add_structs(contract_infos: dict[str, ContractInfo], structs: StructInfo | list[StructInfo]):
+    """_summary_
+
+    Parameters
+    ----------
+    structs : StructInfo | list[StructInfo]
+        _description_
+    """
+    if not isinstance(structs, list):
+        structs = [structs]
+    for struct in structs:
+        info = contract_infos.get(struct.contract_name)
+        if info:
+            info.structs[struct.name] = struct
+        else:
+            contract_infos[struct.contract_name] = ContractInfo(
+                abi=[],
+                bytecode="",
+                contract_name=struct.contract_name,
+                structs={struct.name: struct},
+                events={},
+            )
+
+
+def add_events(contract_infos: dict[str, ContractInfo], events: EventInfo | list[EventInfo], contract_name: str):
+    """_summary_
+
+    Parameters
+    ----------
+    events : EventInfo | list[EventInfo]
+        _description_
+    contract_name : str
+        _description_
+    """
+    if not isinstance(events, list):
```

Comment:
does it make sense to check for Sequence instead of list? i.e. can one input a tuple?